src/include/index/ivf_pq_index.h

```diff
@@ -0,0 +1,1576 @@
+      auto sub_end = (subspace + 1) * dimension_ / num_subspaces_;
+
+      auto local_sub_distance = SubDistance{sub_begin, sub_end};
+
+      // @todo Make choice of kmeans init configurable
+      sub_kmeans_random_init(
+          training_set, cluster_centroids_, sub_begin, sub_end, 0xdeadbeef);
+
+      // sub_kmeans will invoke the sub_distance function with centroids
+      // against new_centroids, and will call flat::qv_partition with centroids
+      // and training_set which will invoke the sub_distance function with
+      // centroids against training_set (though that can perhaps be reversed,
+      // but will have to make sure asymmetric distance gets passed in).
+      // operator()() is a function template, so it should do the "right thing"
+      // @note we are doing this for one subspace at a time
+      auto&& [iters, conv] = sub_kmeans<
+          std::remove_cvref_t<decltype(training_set)>,
+          std::remove_cvref_t<decltype(cluster_centroids_)>,
+          SubDistance>(
+          training_set,
+          cluster_centroids_,
+          sub_begin,
+          sub_end,
+          num_clusters_,
+          tol_,
+          max_iter_,
+          num_threads_);
+
+      max_local_iters_taken = std::max(max_local_iters_taken, iters);
+      min_local_conv = std::min(min_local_conv, conv);
+    }
+
+    // Create tables of distances storing distance between encoding keys,
+    // one table for each subspace.  That is, distance_tables_[i](j, k) is
+    // the distance between the jth and kth centroids in the ith subspace.
+    // The distance between two encoded vectors is looked up using the
+    // keys of the vectors in each subspace (summing up the results obtained
+    // from each subspace).
+    // @todo SIMDize with subspace iteration in inner loop
+    for (size_t subspace = 0; subspace < num_subspaces_; ++subspace) {
+      auto sub_begin = subspace * sub_dimension_;
+      auto sub_end = (subspace + 1) * sub_dimension_;
+      auto local_sub_distance = SubDistance{sub_begin, sub_end};
+
+      for (size_t i = 0; i < num_clusters_; ++i) {
+        for (size_t j = 0; j < num_clusters_; ++j) {
+          auto sub_distance =
+              local_sub_distance(cluster_centroids_[i], cluster_centroids_[j]);
+          distance_tables_[subspace](i, j) = sub_distance;
+        }
+      }
+    }
+
+    return std::make_tuple(max_local_iters_taken, min_local_conv);
+  }
+
+  /***************************************************************************
+   *
+   * Distance functions for pq encoded vectors
+   *
+   ***************************************************************************/
+
+  // @todo IMPORTANT: We need to do some abstraction penalty tests to make sure
+  // that the distance functions are inlined.
+  // @todo Make this SIMD friendly -- do multiple subspaces at a time
+  // For each (i, j), distances should be stored contiguously
+  float sub_distance_symmetric(auto&& a, auto&& b) const {
```

Comment:
We should make the difference between `symmetric` and `asymmetric` distances more clear.

From my understanding `symmetric` means that both queries are PQ embedded and `asymmetric` that `b` is embedded and `a` is not. We also only use the `asymmetric` distance atm however we still generate `distance_tables_` for all cases even if it is only used for `asymmetric` distance computation.

Reply:
That is also my understanding of `symmetric` vs `asymmetric` distance.  The index is created to be agnostic to which query might be presented to it, so it computes the `distance_tables` regardless.  It was also intended to have both `symmetric` and `asymmetric` queries, like `flat_pq_index` but I decided to just put in queries to match names/functionality from `ivf_flat` and picked `asymmetric` as the underlying query.  

We should add support for `symmetric` queries.   If we have the options of finite vs infinite (and maybe even finer grained selection of algorithm) as well as `symmetric` vs `asymmetric` we may want to decide on a clean API -- though `query_{finite, infinite}_ram_{symmetric,asymmetric}` is probably as good a naming scheme as any.  I can make this change or we can leave it to another PR.  Besides the functions we will also need additional unit tests.

We could conceivably only create (or load) the distance tables the first time a symmetric query is requested.  But my own opinion is that we should have the index complete when it is constructed.  If this change is to be made, it should probably be its own PR.

Reply:
Lets add some more comments for symmetric vs asymmetric in this PR and some TODOs and tasks to follow up on supporting `symmetric` queries. Curious, @lums658 which method do you expect to be the most performant? Shouldn't it be `symmetric` with using the distance table?

Reply:
For just the distance function, if you are given an already-encoded query, symmetric is probably faster.  However, if you are given an unencoded query vector, encoding the query vector is quite expensive.  I haven't done the experiments to see if that would be amortized over an entire query or not.

OTOH, we can accelerate encoding by quite a bit by doing it with SIMD instructions, so that will definitely change the comparison.

Story 45153 created for this.